File: 2015/go/day14/day14.go
package day14

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

//Day14 Reindeer Olympics
func Day14(inputFile string) error {
	input, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return err
	}

	fmt.Println("Day 14: Reindeer Olympics")
	part1, err := part1(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 1: %d\n", part1)

	part2, err := part2(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 2: %d\n", part2)

	return nil
}

var reindeerRE = regexp.MustCompile(`([A-Za-z]+) can fly (\d+) km/s for (\d+) seconds, but then must rest for (\d+) seconds.`)

type reindeer struct {
	velocity int
	fly      int
	rest     int
}

func parseReindeer(input string) (map[string]reindeer, error) {
	deer := make(map[string]reindeer)
	for _, line := range strings.Split(input, "\n") {
		matches := reindeerRE.FindStringSubmatch(line)
		if len(matches) < 5 {
			return nil, fmt.Errorf("Invalid input: %s", line)
		}

		velocity, err := strconv.Atoi(matches[2])
		if err != nil {
			return nil, err
		}
		fly, err := strconv.Atoi(matches[3])
		if err != nil {
			return nil, err
		}
		rest, err := strconv.Atoi(matches[4])
		if err != nil {
			return nil, err
		}
		deer[matches[1]] = reindeer{
			velocity: velocity,
			fly:      fly,
			rest:     rest,
		}
	}
	return deer, nil
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func part1(input string) (int, error) {
	deer, err := parseReindeer(input)
	if err != nil {
		return 0, err
	}

	dists := make([]int, 0, len(deer))
	for _, r := range deer {
		dists = append(dists, (2503/(r.fly+r.rest))*(r.velocity*r.fly)+min(2503%(r.fly+r.rest), r.fly)*r.velocity)
	}

	sort.Sort(sort.Reverse(sort.IntSlice(dists)))
	return dists[0], nil
}

func part2(input string) (int, error) {
	deer, err := parseReindeer(input)
	if err != nil {
		return 0, err
	}
	scores := make(map[string]int)

	for i := 1; i <= 2503; i++ {
		max := 0
		var winners []string
		for n, r := range deer {
			score := (i/(r.fly+r.rest))*(r.velocity*r.fly) + min(i%(r.fly+r.rest), r.fly)*r.velocity
			if score > max {
				winners = nil
				winners = append(winners, n)
				max = score
			} else if score == max {
				winners = append(winners, n)
			}
		}

		for _, n := range winners {
			scores[n]++
		}
	}

	sc := make([]int, 0, len(scores))
	for _, v := range scores {
		sc = append(sc, v)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(sc)))
	return sc[0], nil
}

File: 2015/rust/src/day14.rs
use lazy_static::lazy_static;
use regex::Regex;
use simple_error::require_with;
use std::cmp::min;
use std::collections::HashMap;
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;
use std::time::Instant;

pub fn day14(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 14: Reindeer Olympics");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    let start = Instant::now();
    println!("\tPart 1: {} ({:?})", part1(&input)?, start.elapsed());
    let second = Instant::now();
    println!("\tPart 2: {} ({:?})", part2(&input)?, second.elapsed());
    println!("\t\t Completed in {:?}", start.elapsed());
    Ok(())
}

lazy_static! {
    static ref REINDEER_RE: Regex = Regex::new(
        r"([A-Za-z]+) can fly (\d+) km/s for (\d+) seconds, but then must rest for (\d+) seconds."
    )
    .unwrap();
}

#[derive(Debug)]
struct Reindeer {
    velocity: i32,
    fly: i32,
    rest: i32,
}

fn parse_reindeer(input: &str) -> Result<HashMap<&str, Reindeer>, Box<dyn Error>> {
    let mut reindeer = HashMap::new();
    for line in input.lines() {
        let caps = require_with!(
            REINDEER_RE.captures(line),
            &format!("Invalid input: {}", line)
        );
        let name = require_with!(caps.get(1), &format!("Invalid input: {}", line)).as_str();
        let deer = Reindeer {
            velocity: require_with!(caps.get(2), &format!("Invalid input: {}", line))
                .as_str()
                .parse::<i32>()?,
            fly: require_with!(caps.get(3), &format!("Invalid input: {}", line))
                .as_str()
                .parse::<i32>()?,
            rest: require_with!(caps.get(4), &format!("Invalid input: {}", line))
                .as_str()
                .parse::<i32>()?,
        };
        reindeer.insert(name, deer);
    }
    Ok(reindeer)
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    let reindeer = parse_reindeer(input)?;
    //println!("{:?}", reindeer);
    Ok(reindeer
        .values()
        .map(|r| {
            (2503 / (r.fly + r.rest)) * (r.velocity * r.fly)
                + min(2503 % (r.fly + r.rest), r.fly) * r.velocity
        })
        .max()
        .unwrap())
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    let reindeer = parse_reindeer(input)?;
    let mut scores: HashMap<&str, i32> = HashMap::new();

    for i in 1..=2503 {
        let mut max = 0;
        let mut winners = vec![];
        for (n, r) in &reindeer {
            let score = (i / (r.fly + r.rest)) * (r.velocity * r.fly)
                + min(i % (r.fly + r.rest), r.fly) * r.velocity;
            if score > max {
                winners.clear();
                winners.push(n);
                max = score;
            } else if score == max {
                winners.push(n);
            }
        }

        for n in winners {
            let score = scores.entry(n).or_insert(0);
            *score += 1;
        }
    }
    Ok(scores.values().map(|x| *x).max().unwrap())
}
